File: sdk/python/nexusweb3/contracts/escrow.py
"""AgentEscrowV2 — milestone escrow (`IAgentEscrowV2`)."""

from __future__ import annotations

from dataclasses import replace

from web3 import Web3

from ..tx import TxResult
from ..types import CreateParams, Job, Milestone, Payout, coerce_bytes32
from .base import Ownable2StepClient

__all__ = ["EscrowClient"]


class EscrowClient(Ownable2StepClient):
    """Client and provider may act through their principals or their operators."""

    # ─── Client ─────────────────────────────────────────────────────────
    def create_job(self, params: CreateParams) -> TxResult:
        """Pull `params.total` from the client principal. Needs a prior USDC approval."""
        result = self._send("createJob", params.to_tuple())
        return self._with_job_id(result)

    def create_job_with_permit(
        self,
        params: CreateParams,
        permit_deadline: int,
        v: int,
        r: bytes,
        s: bytes,
    ) -> TxResult:
        """Create a job using an EIP-2612 signature instead of a prior approval.

        The contract calls `permit` inside a try/catch, so a token without EIP-2612 support
        fails later on `transferFrom` rather than here. Use :func:`nexusweb3.permit.sign_permit`.
        """
        result = self._send(
            "createJobWithPermit", params.to_tuple(), int(permit_deadline), int(v), bytes(r), bytes(s)
        )
        return self._with_job_id(result)

    def approve_milestone(self, job_id: int, index: int) -> TxResult:
        """Pay the provider for milestone `index` (minus fee) and advance the job.

        The returned :class:`TxResult` carries `payouts` decoded from `PayoutSettled`.
        """
        return self._with_payouts(self._send("approveMilestone", int(job_id), int(index)))

    def reject_milestone(self, job_id: int, index: int, reason_hash: str | bytes | None = None) -> TxResult:
        """Send a Submitted milestone back, inside `REVIEW_WINDOW` and at most `MAX_REJECTIONS` times."""
        return self._send("rejectMilestone", int(job_id), int(index), coerce_bytes32(reason_hash))

    def cancel_job(self, job_id: int) -> TxResult:
        """Full refund. Allowed before acceptance, and after it only while nothing was ever submitted.

        The returned :class:`TxResult` carries `payouts` decoded from `PayoutSettled`.
        """
        return self._with_payouts(self._send("cancelJob", int(job_id)))

    # ─── Provider ───────────────────────────────────────────────────────
    def accept_job(self, job_id: int) -> TxResult:
        """Bind the provider to the offer.

        A created job is only an offer: `submit_milestone`, `approve_milestone` and `dispute` all
        revert with `NotAccepted` until this lands, and the client may cancel for a full refund.
        Must happen before the job deadline, and only once (`AlreadyAccepted` afterwards).
        """
        return self._send("acceptJob", int(job_id))

    def submit_milestone(self, job_id: int, index: int, deliverable_hash: str | bytes) -> TxResult:
        return self._send("submitMilestone", int(job_id), int(index), coerce_bytes32(deliverable_hash))

    def claim_approval(self, job_id: int, index: int) -> TxResult:
        """Provider self-approval of a milestone the client left Submitted past `REVIEW_WINDOW`.

        The returned :class:`TxResult` carries `payouts` decoded from `PayoutSettled`.
        """
        return self._with_payouts(self._send("claimApproval", int(job_id), int(index)))

    # ─── Either party / arbiter / anyone ────────────────────────────────
    def dispute(self, job_id: int, reason_hash: str | bytes | None = None) -> TxResult:
        return self._send("dispute", int(job_id), coerce_bytes32(reason_hash))

    def resolve(self, job_id: int, provider_bps: int) -> TxResult:
        """Arbiter only: split the remaining funds, `provider_bps` out of 10_000 to the provider.

        The returned :class:`TxResult` carries `payouts` decoded from `PayoutSettled`.
        """
        return self._with_payouts(self._send("resolve", int(job_id), int(provider_bps)))

    def settle_expired(self, job_id: int) -> TxResult:
        """Anyone: close out a job time has decided.

        Open jobs qualify past `expiry_of`, Disputed ones `DISPUTE_GRACE` after `disputed_at`.
        Every Submitted milestone vests to the provider and every Pending one refunds the client,
        so client silence no longer claws back delivered work. The returned :class:`TxResult`
        carries `to_provider` and `to_client` decoded from `JobExpired`, plus `payouts`.
        """
        result = self._with_payouts(self._send("settleExpired", int(job_id)))
        args = self._event_args("JobExpired", result.receipt)
        if not args:
            return result
        return replace(
            result, to_provider=int(args[0]["toProvider"]), to_client=int(args[0]["toClient"])
        )

    def withdraw_claimable(self, account: str, to: str) -> TxResult:
        """Sweep funds parked for `account` after a failed payout into `to`.

        Signed by `account` itself or one of its operators; `to` must not be the zero address.
        The returned :class:`TxResult` carries `withdrawn` decoded from `ClaimableWithdrawn`.
        """
        result = self._send(
            "withdrawClaimable", Web3.to_checksum_address(account), Web3.to_checksum_address(to)
        )
        args = self._event_args("ClaimableWithdrawn", result.receipt)
        if not args:
            return result
        return replace(result, withdrawn=int(args[0]["amount"]))

    # ─── Views ──────────────────────────────────────────────────────────
    def get_job(self, job_id: int) -> Job:
        return Job.from_tuple(self._call("getJob", int(job_id)))

    def get_milestones(self, job_id: int) -> list[Milestone]:
        return [Milestone.from_tuple(row) for row in self._call("getMilestones", int(job_id))]

    def get_jobs_of(self, account: str, offset: int = 0, limit: int = 50) -> list[int]:
        rows = self._call("getJobsOf", Web3.to_checksum_address(account), int(offset), int(limit))
        return [int(job_id) for job_id in rows]

    def job_count_of(self, account: str) -> int:
        return int(self._call("jobCountOf", Web3.to_checksum_address(account)))

    def job_count(self) -> int:
        return int(self._call("jobCount"))

    def expiry_of(self, job_id: int) -> int:
        """Effective expiry: `max(deadline, latest submission + REVIEW_WINDOW)`.

        The moment an Open job becomes settleable with :meth:`settle_expired`, and the moment
        `dispute` stops working.
        """
        return int(self._call("expiryOf", int(job_id)))

    def review_window(self) -> int:
        """Seconds a client may leave a submitted milestone unreviewed before the provider can claim."""
        return int(self._call("REVIEW_WINDOW"))

    def dispute_grace(self) -> int:
        """Seconds a Disputed job waits for its arbiter before anyone may :meth:`settle_expired` it."""
        return int(self._call("DISPUTE_GRACE"))

    def max_rejections(self) -> int:
        """Rejections one milestone tolerates before resubmission reverts with `TooManyRejections`."""
        return int(self._call("MAX_REJECTIONS"))

    def max_reputation_per_pair(self) -> int:
        """Reputation entries a single client/provider pair may generate, so a pair cannot farm score."""
        return int(self._call("MAX_REPUTATION_PER_PAIR"))

    def min_reputation_value(self) -> int:
        """Settlement value below which no reputation is written, so dust jobs cannot farm score."""
        return int(self._call("MIN_REPUTATION_VALUE"))

    def claimable(self, account: str) -> int:
        return int(self._call("claimable", Web3.to_checksum_address(account)))

    def fee_bps(self) -> int:
        return int(self._call("feeBps"))

    def payment_token(self) -> str:
        return str(self._call("paymentToken"))

    def reputation(self) -> str:
        return str(self._call("reputation"))

    def audit_log(self) -> str:
        return str(self._call("auditLog"))

    def kill_switch(self) -> str:
        return str(self._call("killSwitch"))

    def fee_router(self) -> str:
        return str(self._call("feeRouter"))

    # ─── Owner ──────────────────────────────────────────────────────────
    def set_fee_bps(self, new_bps: int) -> TxResult:
        return self._send("setFeeBps", int(new_bps))

    def set_modules(self, reputation: str, audit_log: str, kill_switch: str, fee_router: str) -> TxResult:
        return self._send(
            "setModules",
            Web3.to_checksum_address(reputation),
            Web3.to_checksum_address(audit_log),
            Web3.to_checksum_address(kill_switch),
            Web3.to_checksum_address(fee_router),
        )

    def _with_payouts(self, result: TxResult) -> TxResult:
        """Attach every `PayoutSettled` this transaction emitted, in log order.

        The event fires for a delivered transfer and for one that bounced into `claimable` alike,
        so a caller can see where the money actually went without a second round trip.
        """
        payouts = tuple(
            Payout.from_args(args) for args in self._event_args("PayoutSettled", result.receipt)
        )
        return replace(result, payouts=payouts)

    def _with_job_id(self, result: TxResult) -> TxResult:
        job_id = self._first_event_arg("JobCreated", result.receipt, "jobId")
        return TxResult(hash=result.hash, receipt=result.receipt, job_id=job_id)

File: sdk/ts/src/modules/escrow.ts
import { getContract, zeroAddress, zeroHash, type Address, type Hex } from 'viem';
import { AgentEscrowV2Abi } from '../abis/AgentEscrowV2.js';
import { toBaseUnits } from '../amount.js';
import { eventBigInt, requireEvent, sendWrite, type Context } from '../internal.js';
import {
  decodeJobStatus,
  decodeMilestoneStatus,
  type CreateJobParams,
  type CreateJobResult,
  type Job,
  type Milestone,
  type PermitSignature,
  type SettleExpiredResult,
  type TxResult,
  type WithdrawClaimableResult,
} from '../types.js';

type RawJob = Readonly<{
  client: Address;
  provider: Address;
  arbiter: Address;
  total: bigint;
  released: bigint;
  refunded: bigint;
  deadline: number;
  createdAt: number;
  acceptedAt: number;
  disputedAt: number;
  milestoneCount: number;
  approvedCount: number;
  everSubmitted: boolean;
  status: number;
  termsHash: Hex;
}>;

type RawMilestone = Readonly<{
  amount: bigint;
  deliverableHash: Hex;
  submittedAt: number;
  rejections: number;
  status: number;
}>;

/** Solidity `CreateParams`, with the SDK's optional fields filled in. */
interface CreateParamsTuple {
  client: Address;
  provider: Address;
  arbiter: Address;
  milestoneAmounts: readonly bigint[];
  deadline: number;
  termsHash: Hex;
}

function toCreateParams(params: CreateJobParams): CreateParamsTuple {
  return {
    client: params.client,
    provider: params.provider,
    arbiter: params.arbiter ?? zeroAddress,
    milestoneAmounts: params.milestoneAmounts.map((amount, index) =>
      toBaseUnits(amount, `milestoneAmounts[${index}]`),
    ),
    deadline: params.deadline,
    termsHash: params.termsHash ?? zeroHash,
  };
}

function decodeJob(job: RawJob): Job {
  const { status, ...rest } = job;
  return { ...rest, status: decodeJobStatus(status) };
}

function decodeMilestone(milestone: RawMilestone): Milestone {
  const { status, ...rest } = milestone;
  return { ...rest, status: decodeMilestoneStatus(status) };
}

export interface EscrowModule {
  readonly address: Address;
  /** Client side: pulls `sum(milestoneAmounts)` of the payment token from `params.client`. */
  createJob(params: CreateJobParams): Promise<CreateJobResult>;
  /** Same, but consumes an EIP-2612 permit first so no prior `approve` is needed. */
  createJobWithPermit(
    params: CreateJobParams,
    permitDeadline: bigint,
    signature: PermitSignature,
  ): Promise<CreateJobResult>;
  /**
   * Provider side: bind the provider to the offer. A job stays a mere offer until this lands —
   * `submitMilestone`, `approveMilestone` and `dispute` all revert with `NotAccepted` before it,
   * and the client may cancel for a full refund. Must happen before the job deadline.
   */
  acceptJob(jobId: bigint): Promise<TxResult>;
  submitMilestone(jobId: bigint, index: number, deliverableHash: Hex): Promise<TxResult>;
  /** Provider side: take a milestone the client left Submitted past the review window. */
  claimApproval(jobId: bigint, index: number): Promise<TxResult>;
  approveMilestone(jobId: bigint, index: number): Promise<TxResult>;
  /** Client side, inside the review window only; at most {@link EscrowModule.maxRejections} per milestone. */
  rejectMilestone(jobId: bigint, index: number, reasonHash?: Hex): Promise<TxResult>;
  /** Client side: full refund. Only before acceptance, or after it while nothing was ever submitted. */
  cancelJob(jobId: bigint): Promise<TxResult>;
  /** Either party; needs an arbiter on the job and an accepted job. */
  dispute(jobId: bigint, reasonHash?: Hex): Promise<TxResult>;
  /** Arbiter only: split the remaining balance, `providerBps` out of 10_000 to the provider. */
  resolve(jobId: bigint, providerBps: number): Promise<TxResult>;
  /**
   * Anyone, once the job expired (Open past {@link EscrowModule.expiryOf}) or a dispute went
   * unresolved for {@link EscrowModule.disputeGrace}. Every Submitted milestone vests to the
   * provider, every Pending one refunds the client.
   */
  settleExpired(jobId: bigint): Promise<SettleExpiredResult>;
  /**
   * Sweep funds parked after a failed payout. Callable by `account` itself or one of its
   * operators; `to` is where the tokens land and must not be the zero address.
   */
  withdrawClaimable(account: Address, to: Address): Promise<WithdrawClaimableResult>;
  getJob(jobId: bigint): Promise<Job>;
  getMilestones(jobId: bigint): Promise<Milestone[]>;
  getJobsOf(account: Address, offset: bigint, limit: bigint): Promise<readonly bigint[]>;
  jobCountOf(account: Address): Promise<bigint>;
  jobCount(): Promise<bigint>;
  /** max(deadline, last submission + review window) — the moment `settleExpired` opens. */
  expiryOf(jobId: bigint): Promise<number>;
  claimable(account: Address): Promise<bigint>;
  feeBps(): Promise<bigint>;
  paymentToken(): Promise<Address>;
  /** Seconds a client may leave a Submitted milestone unreviewed before the provider can claim it. */
  reviewWindow(): Promise<bigint>;
  /** Seconds a Disputed job waits for its arbiter before anyone may `settleExpired` it. */
  disputeGrace(): Promise<bigint>;
  /** Rejections a single milestone tolerates before resubmission is refused. */
  maxRejections(): Promise<number>;
  /** Reputation entries one client/provider pair can generate, so a pair cannot farm score. */
  maxReputationPerPair(): Promise<bigint>;
  /** Settlement value below which no reputation is written at all, so dust jobs cannot farm score. */
  minReputationValue(): Promise<bigint>;
}

export function createEscrowModule(ctx: Context): EscrowModule {
  const address = ctx.addresses.escrow;
  const reader = getContract({ address, abi: AgentEscrowV2Abi, client: ctx.publicClient });

  const withJobId = (result: TxResult): CreateJobResult => {
    const event = requireEvent(AgentEscrowV2Abi, 'JobCreated', address, result.receipt.logs);
    return { ...result, jobId: eventBigInt(event.args, 'jobId') };
  };

  return {
    address,
    async createJob(params) {
      return withJobId(await sendWrite(ctx, address, AgentEscrowV2Abi, 'createJob', [toCreateParams(params)]));
    },
    async createJobWithPermit(params, permitDeadline, signature) {
      const result = await sendWrite(ctx, address, AgentEscrowV2Abi, 'createJobWithPermit', [
        toCreateParams(params),
        permitDeadline,
        signature.v,
        signature.r,
        signature.s,
      ]);
      return withJobId(result);
    },
    async acceptJob(jobId) {
      const result = await sendWrite(ctx, address, AgentEscrowV2Abi, 'acceptJob', [jobId]);
      // Proves the ABI and the deployed contract agree on the lifecycle, same as createJob.
      requireEvent(AgentEscrowV2Abi, 'JobAccepted', address, result.receipt.logs);
      return result;
    },
    async submitMilestone(jobId, index, deliverableHash) {
      return sendWrite(ctx, address, AgentEscrowV2Abi, 'submitMilestone', [jobId, index, deliverableHash]);
    },
    async claimApproval(jobId, index) {
      return sendWrite(ctx, address, AgentEscrowV2Abi, 'claimApproval', [jobId, index]);
    },
    async approveMilestone(jobId, index) {
      return sendWrite(ctx, address, AgentEscrowV2Abi, 'approveMilestone', [jobId, index]);
    },
    async rejectMilestone(jobId, index, reasonHash = zeroHash) {
      return sendWrite(ctx, address, AgentEscrowV2Abi, 'rejectMilestone', [jobId, index, reasonHash]);
    },
    async cancelJob(jobId) {
      return sendWrite(ctx, address, AgentEscrowV2Abi, 'cancelJob', [jobId]);
    },
    async dispute(jobId, reasonHash = zeroHash) {
      return sendWrite(ctx, address, AgentEscrowV2Abi, 'dispute', [jobId, reasonHash]);
    },
    async resolve(jobId, providerBps) {
      return sendWrite(ctx, address, AgentEscrowV2Abi, 'resolve', [jobId, providerBps]);
    },
    async settleExpired(jobId) {
      const result = await sendWrite(ctx, address, AgentEscrowV2Abi, 'settleExpired', [jobId]);
      const event = requireEvent(AgentEscrowV2Abi, 'JobExpired', address, result.receipt.logs);
      return {
        ...result,
        toProvider: eventBigInt(event.args, 'toProvider'),
        toClient: eventBigInt(event.args, 'toClient'),
      };
    },
    async withdrawClaimable(account, to) {
      const result = await sendWrite(ctx, address, AgentEscrowV2Abi, 'withdrawClaimable', [account, to]);
      const event = requireEvent(AgentEscrowV2Abi, 'ClaimableWithdrawn', address, result.receipt.logs);
      return { ...result, amount: eventBigInt(event.args, 'amount') };
    },
    async getJob(jobId) {
      return decodeJob(await reader.read.getJob([jobId]));
    },
    async getMilestones(jobId) {
      const milestones = await reader.read.getMilestones([jobId]);
      return milestones.map(decodeMilestone);
    },
    getJobsOf: (account, offset, limit) => reader.read.getJobsOf([account, offset, limit]),
    jobCountOf: (account) => reader.read.jobCountOf([account]),
    jobCount: () => reader.read.jobCount(),
    expiryOf: (jobId) => reader.read.expiryOf([jobId]),
    claimable: (account) => reader.read.claimable([account]),
    feeBps: () => reader.read.feeBps(),
    paymentToken: () => reader.read.paymentToken(),
    reviewWindow: () => reader.read.REVIEW_WINDOW(),
    disputeGrace: () => reader.read.DISPUTE_GRACE(),
    maxRejections: () => reader.read.MAX_REJECTIONS(),
    maxReputationPerPair: () => reader.read.MAX_REPUTATION_PER_PAIR(),
    minReputationValue: () => reader.read.MIN_REPUTATION_VALUE(),
  };
}
